fpga_interchange/arch_pack_clusters.cc

```diff
@@ -370,6 +506,344 @@ static bool check_cluster_cells_compatibility(CellInfo *old_cell, CellInfo *new_
+                log_info("Have: %s %d\n", cell_type.c_str(ctx), counter[cell_type]);
+            if(!cells_in_macro.count(cell_type) || !counter.count(cell_type) || cells_in_macro[cell_type] != counter[cell_type])
+                failed = true;
+            if(failed && ctx->verbose)
+                log_info("Cell count stage failed, for sure not this cluster\n");
+            if(failed)
+                break;
+        }
+        if(failed){
+            roots[ci->macro_parent] = nullptr;
+            continue;
+        }
+
+        // Arc consistency
+        dict<uint32_t, pool<CellInfo *, hash_ptr_ops>> idx_to_cells;
+        // First singular constraints, like used cell type and used_cell ports
+        for (auto &cell : macro_to_cells[ci->macro_parent])
+            for (auto &node : cluster->connection_graph)
+                if (IdString(node.cell_type) == cell->type)
+                    if (node.idx != 0 && cell->name != ci->name ||
+                        node.idx == 0 && cell->name == ci->name ){
+                        idx_to_cells[node.idx].insert(cell);
+                    }
+
+        for (auto &arc : idx_to_cells){
+            std::vector<CellInfo *> remove_cell;
+            pool<IdString> used_ports;
+            for (const auto &port : cluster->connection_graph[arc.first].used_ports)
+                used_ports.insert(IdString(port.name));
+            for (const auto &cell : arc.second){
+                uint32_t count = 0;
+                for (const auto &port : cell->ports){
+                    if (!used_ports.count(port.first)){
+                        remove_cell.push_back(cell);
+                        break;
+                    }
+                    count++;
+                }
+                if (count != used_ports.size()){
+                    remove_cell.push_back(cell);
+                    break;
+                }
+            }
+            for (const auto &cell : remove_cell){
+                arc.second.erase(cell);
+            }
+        }
+        if (ctx->debug){
+            log_info("After mono constraints are applied\n");
+            dict<CellInfo *, pool<uint32_t>, hash_ptr_ops> possible_idx;
+            for (const auto &arc : idx_to_cells)
+                for (const auto &cell : arc.second)
+                    possible_idx[cell].insert(arc.first);
+
+            for (const auto arc : possible_idx){
+                log_info("Possible idx %s:\n", arc.first->name.c_str(ctx));
+                for (const auto idx : arc.second)
+                    log_info("    - %d\n", idx);
+            }
+        }
+        // Solve for binary constraints
+        std::queue<std::pair<uint32_t, uint32_t>> workqueue;
+        for (const auto &arc : idx_to_cells)
+            for (const auto &connection : cluster->connection_graph[arc.first].connections)
+                workqueue.push(std::pair<uint32_t, uint32_t>(arc.first, connection.target_idx));
```

Comment:
I think this can be `emplace` instead of `push`

Reply:
Changed to emplace